docs/proposals/ktl.md

```diff
@@ -0,0 +1,243 @@
+    for a given resource.
+  - generic commands that are agnostic to specific APIs, but instead discover API metadata and work against
+    all APIs following published conventions.  e.g. a `scale` that works against anything with a scale sub resource
+
+
+### Library infrastructure
+
+Develop shared set of client / cli libraries that handle registering flags / environment variables / config files
+and provide common functionality to cli clients of Kubernetes.
+
+- reading and parsing kubeconfig
+- printing libraries
+- reading and parsing config into objects from files, directories, stdin
+- indexing + querying discovery API and openapi data
+- manipulating unstructured objects using openapi schema
+- merging objects
+- generating patches from objects
+- writing objects to the server - optimistic locking failure retries with exp backoff
+- providing tty for subresources that accept - e.g. exec, attach
+- defining shared exit codes (e.g. exit code when not doing a conditional write)
+- test infrastructure for unit and integration tests
+  - support version skew testing
+
+### Build and release infrastructure
+
+Develop build triggers to automatically cut and publish builds based on the presence of GitHub tags.  Aggregate
+release notes from vendored commands.
+
+- use GCP container builder + mirrored GCP source repo
+- publish binary to gs:// bucket
+- publish binary + release notes to GitHub releases page
+
+## Graduation Criteria
+
+None
+
+## Alternatives [optional]
+
+### Keeping the kubectl name instead of rebranding
+
+Alternatively we could call the new command `kubectl` and attach all of the legacy kubectl commands
+at the root level.  This would make it look and feel exactly like `kubectl`, but allow for new pieces
+to be built out of kubernetes/kubernetes.
+
+Steps:
+
+- rename `kubernetes/kubernets/cmd/kubectl` to `kubernetes/kubernets/cmd/legacykubectl`
+- rename `kubernetes/kubernets/pkg/kubectl` to `kubernetes/kubernets/pkg/legacykubectl`
+- create new `kubernetes/kubectl/cmd/kubectl` command
+- vendor kubectl subcommands directly under `kubernetes/kubectl/cmd/kubectl` root command
+
+Tradeoffs:
+
+Keeping kubectl as name: Need to make it continue to look and act like kubectl
+
+- Minimal change from users perspective
+- Can easily swap with existing kubectl without updating docs, blog posts, etc
+
+Renaming to ktl: Need to get everyone to use the new command
+
+- Easier to phase out old commands with new ones while changing behavior
+- Easier to restructure command layout
+- Easier to redefine command wide things - such as version skew support
+
+
+### Don't vendor in commands, make them plugins instead*
+
+This would be a more complicated approach that can be considered in a later iteration.
+
+## Unresolved Questions [optional]
+
+- Should we rename / restructure the command, or keep kubectl and all the bagage that comes with?
```

Comment:
The question is what CLI API guarantees *must* we preserve (which ones will users accept, and which will users reject)?  Do we preserve the scripting semantics of get?  The resource argument ordering?  Places where the CLI invocation is broken is where we don't want to inherit, places where the cost to end users is going to be very high to switch is where we do.